File: 백준/Gold/22862. 가장 긴 짝수 연속한 부분 수열 （large）/가장 긴 짝수 연속한 부분 수열 （large）.cc
#include <bits/stdc++.h>

using namespace std;

int n, k;
int arr[1000001];

int main()
{
	cin >> n >> k;

	for (int i = 0; i < n; i++)
	{
		cin >> arr[i];
	}

	int st = 0;
	int en = 0;
	int cnt = (arr[st] & 1) ? 1 : 0; // 현재 구간의 홀수 개수
	int Max = 0;

	while (1)
	{
		while (en < n - 1) // en을 뒤로 이동시킬 수 이씅ㄹ 경우
		{
			if (arr[en + 1] & 1) // en의 다음 요소가 홀수인 경우
			{
				if (cnt < k) // 홀수를 삭제할 수 있는 경우,
				{
					cnt++;
				}
				else // 홀수를 삭제할 수 없는 경우
				{
					break;
				}
			}

			en++;
		}
		
		if (st >= n || en >= n) // 범위를 벗어나면
			break;

		Max = max(Max, en - st + 1 - cnt);

		if (arr[st] & 1)
			cnt--;
		
		st++;

	}

	cout << Max;

	return 0;
}

File: 백준/Gold/22862. 가장 긴 짝수 연속한 부분 수열 （large）/가장 긴 짝수 연속한 부분 수열 （large）.java
import java.util.*;
import java.io.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    static int n, k, max = Integer.MIN_VALUE;
    static int[] arr;

    public static void main(String[] args) throws Exception {
        st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());

        arr = new int[n];

        st = new StringTokenizer(br.readLine());

        for (int i = 0 ; i < n ; i ++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        int left = 0;
        int right = 0;
        int len = 0;
        int oddCount = 0;

        while(left < n || right < n) {

            // 짝수면
            if (right < n && arr[right] % 2 == 0) {
                len ++;
                right ++;
            }
            // 홀수인데
            else {
                // 삭제할 수 있다면
                if (oddCount < k) {
                    oddCount ++;
                    right ++;
                }
                // 삭제할 수 없으면
                else {
                    // left 인덱스 값이 홀수인지 짝수인지 확인

                    // 짝수면
                    if (left < n && arr[left] % 2 == 0) {
                        len --;
                    }
                    // 홀수면
                    if (left < n && arr[left] % 2 == 1) {
                        oddCount--;
                    }

                    left ++;
                }
            }

            if (len > max) max = len;
        }

        System.out.println(max);
    }
}